pkg/limits/ingest_limits.go

```diff
@@ -397,104 +338,34 @@ func (s *IngestLimits) evictOldStreamsPeriodic(ctx context.Context) {
 		case <-ctx.Done():
 			return
 		case <-ticker.C:
-			s.evictOldStreams(ctx)
+			cutoff := time.Now().Add(-s.cfg.WindowSize).UnixNano()
+			s.metadata.Evict(cutoff)
 		}
 	}
 }
 
 // updateMetadata updates the metadata map with the provided StreamMetadata.
 // It uses the provided lastSeenAt timestamp as the last seen time.
 func (s *IngestLimits) updateMetadata(rec *logproto.StreamMetadata, tenant string, partition int32, lastSeenAt time.Time) {
-	s.mtx.Lock()
-	defer s.mtx.Unlock()
-
-	// Initialize tenant map if it doesn't exist
-	if _, ok := s.metadata[tenant]; !ok {
-		s.metadata[tenant] = make(map[int32][]streamMetadata)
-	}
-
-	if s.metadata[tenant][partition] == nil {
-		s.metadata[tenant][partition] = make([]streamMetadata, 0)
-	}
+	var (
+		// Use the provided lastSeenAt timestamp as the last seen time
+		recordTime = lastSeenAt.UnixNano()
+		// Get the bucket for this timestamp using the configured interval duration
+		bucketStart = lastSeenAt.Truncate(s.cfg.BucketDuration).UnixNano()
+		// Calculate the rate window cutoff for cleaning up old buckets
+		rateWindowCutoff = lastSeenAt.Add(-s.cfg.RateWindow).UnixNano()
+		// Calculate the total size of the stream
+		totalSize = rec.EntriesSize + rec.StructuredMetadataSize
+	)
 
-	// Partition not assigned to this instance, evict stream
+	evict := false
```

Comment:
I suspect we can delete this, it will make the code simpler too. Instead, just return if partition is not assigned.